episodes/02-data-structures.Rmd

```diff
@@ -0,0 +1,234 @@
+
+```{r remove-na}
+with_na <- c(1, 2, 1, 1, NA, 3, NA ) # vector including missing value
+```
+
+First, let's try to calculate mean for the values in this vector
+```{r remove-na1}
+mean(with_na) # mean() function cannot interpret the missing values
+
+mean(with_na, na.rm = T) # You can add the argument na.rm=TRUE to calculate the result while ignoring the missing values.
+```
+
+However, sometimes, you would like to have the `NA` 
+permanently removed from your vector. 
+For this you need to identify which elements of the vector hold missing values 
+with `is.na()` function. 
+
+```{r remove-na2}
+is.na(with_na) #  This will produce a vector of logical values, stating if a statement 'This element of the vector is a missing value' is true or not
+
+!is.na(with_na) # # The ! operator means negation ,i.e. not is.na(with_na)
+
+```
+
+We know which elements in the vectors are `NA`.
+Now we need to retrieve the subset of the `with_na` vector that is not `NA`.
+Sub-setting in `R` is done with square brackets`[ ]`. 
+
+```{r remove-na3}
+
+without_na <- with_na[ !is.na(with_na) ] # this notation will return only the elements that have TRUE on their respective positions
+
+without_na
+
+```
+
+
+## Factors 
+
+Another important data structure is called a **factor**. 
+Factors look like character data, but are used to represent categorical information.
+
+Factors create a structured relation between the different levels (values) of a
+categorical variable, such as days of the week or responses to a question in a
+survey. While factors look (and often behave) like character vectors, they are 
+actually treated as numbers by `R`. 
+So you need to be very careful when treating them as strings.
+
+### Create factors
+Once created, factors can only contain a pre-defined set of values, 
+known as levels. 
+
+```{r factor-create}
+
+nordic_str <- c('Norway', 'Sweden', 'Norway', 'Denmark', 'Sweden')
+nordic_str # regular character vectors printed out
+
+nordic_cat <- factor(nordic_str) # factor() function converts a vector to factor data type
+nordic_cat # With factors, R prints out additional information - 'Levels'
+
+```
+
+### Inspect factors
+R will treat each unique value from a factor vector as a **level** and (silently)
+assign numerical values to it. 
+This can come in handy when performing statistical analysis. 
+You can inspect and adapt levels of the factor. 
+
+```{r factor-inspect}
+levels(nordic_cat) # returns all levels of a factor vector.  
+
+nlevels(nordic_cat) # returns number of levels in a vector
+```
+
+### Reorder levels
+Note that `R` sorts the levels in the alphabetic order, 
+not in the order of occurrence in the vector. `R` assigns value of:
+
+- 1 to level 'Denmark',
+- 2 to 'Norway' 
+- 3 to 'Sweden'.
+
+This is important as it can affect e.g. the order in which categories are 
+displayed in a plot or which category is taken as a baseline in a statistical model.
+
+You can reorder the categories using `factor()` function.
```

Comment:
```suggestion
You can reorder the categories using `factor()` function. This can be useful, for instance, to select a reference category (first level) in a regression model or for ordering legend items in a plot, rather than using the default category systematically (i.e. based on alphabetical order).
```